src/Build/Graph/ProjectGraph.cs

```diff
@@ -243,6 +297,92 @@ private static ImmutableList<string> ExpandDefaultTargets(ProjectInstance projec
+            string setTargetFrameworkString = projectReference.GetMetadataValue(SetTargetFrameworkMetadataName);
+
+            // The properties on the project reference supercede the ones from the MSBuild task instad of appending.
+            if (string.IsNullOrEmpty(propertiesString)
+                && (!string.IsNullOrEmpty(setConfigurationString) || !string.IsNullOrEmpty(setPlatformString) || !string.IsNullOrEmpty(setTargetFrameworkString)))
+            {
+                propertiesString = $"{setConfigurationString};{setPlatformString};{setTargetFrameworkString}";
+            }
+
+            // If none of these are set, we can just reuse the requestor's global properties directly.
+            if (string.IsNullOrEmpty(propertiesString)
+                && string.IsNullOrEmpty(additionalPropertiesString)
+                && string.IsNullOrEmpty(undefinePropertiesString)
+                && string.IsNullOrEmpty(globalPropertiesToRemoveString))
+            {
+                return requesterGlobalProperties;
+            }
+
+            // Make a copy to avoid mutating the requester
+            var globalProperties = new PropertyDictionary<ProjectPropertyInstance>(requesterGlobalProperties);
+
+            // Append and remove properties as specified by the various metadata
+            MergeIntoPropertyDictionary(globalProperties, propertiesString, PropertiesMetadataName);
+            MergeIntoPropertyDictionary(globalProperties, additionalPropertiesString, AdditionalPropertiesMetadataName);
+            RemoveFromPropertyDictionary(globalProperties, globalPropertiesToRemoveString);
+            RemoveFromPropertyDictionary(globalProperties, undefinePropertiesString);
+
+            return globalProperties;
+        }
+
+        private static void MergeIntoPropertyDictionary(
+            PropertyDictionary<ProjectPropertyInstance> properties,
+            string propertyNameAndValuesString,
+            string syntaxName)
+        {
+            if (!string.IsNullOrEmpty(propertyNameAndValuesString))
+            {
+                if (PropertyParser.GetTableWithEscaping(
+                    null,
+                    null,
+                    null,
+                    propertyNameAndValuesString.Split(PropertySeparator, StringSplitOptions.RemoveEmptyEntries),
+                    out Dictionary<string, string> propertiesTable))
+                {
+                    foreach (KeyValuePair<string, string> pair in propertiesTable)
+                    {
+                        properties[pair.Key] = ProjectPropertyInstance.Create(pair.Key, pair.Value);
+                    }
+                }
+                else
+                {
+                    throw new InvalidProjectFileException(string.Format(CultureInfo.InvariantCulture, ResourceUtilities.GetResourceString("General.InvalidPropertyError"), syntaxName, propertyNameAndValuesString));
```

Comment:
This isn't quite right. This error is logged to the task logger (if provided) in `PropertyParser.GetTableWithEscaping`, but the error is for a specific property/name pair and not the entire string. So the error string is a bit off here.

Reply:
I'm not sure I understand, can you paste the result of an error?

Reply:
Yea, that was super unclear :)

So `PropertyParser.GetTableWithEscaping` takes a task log helper and proceeds to process each split one by one. So like "A=B;C=D" would get processed one by one "A=B" and "C=D". It does have to process them all at once though in the special case when it has to merge entries, like "A=B;C" would get split to "A=B" and "C", but then merged back together such that "A" would get set to "B;C". When it does see something invalid though, it errors on that specific entry, not the entire input. So something like "A;B=C" would split to "A" and "B=C" and then there would be an error about "A" (not "A;B=C"). Something like "A=B;=;=;=" I would guess would be similarly invalid, but the error just mention "=" as the bad string.

Here's the actual resource string in case it helps:
> MSB3100: Syntax for "{0}" parameter is not valid ({1}). Correct syntax is {0}="&lt;name&gt;=&lt;value&gt;".

TL;DR, the error message is for a specific entry, not the entire string.